### PR TITLE
feat(tfidf): add TF-IDF vectorizer

### LINKED ISSUE #104: feat: Add TF-IDF vectorizer
## Summary

Add a standalone TF-IDF (Term Frequency-Inverse Document Frequency) vectorizer. The gem already has `word_hash` for term frequency, but lacks IDF weighting. TF-IDF is the foundation for most classic text classification approaches.

## Motivation

- **Foundation for other classifiers** - Logistic Regression (#102) and kNN (#103) benefit from TF-IDF
- **Better feature weighting** - IDF downweights common words, upweights discriminative terms
- **Industry standard** - TF-IDF is the baseline representation for text ML
- **Improves LSI** - TF-IDF weighted term-document matrix often gives better LSI results
- **Standalone utility** - useful for search, similarity, and feature extraction beyond classification

## Proposed API

```ruby
# Build vocabulary and compute IDF from corpus
vectorizer = Classifier::TFIDF.new
vectorizer.fit(["Dogs are great pets", "Cats are independent", ...])

# Transform text to TF-IDF vector
vector = vectorizer.transform("Dogs are loyal")
# => {dog: 0.7, loyal: 0.5, ...}

# Fit and transform in one step
vectorizer.fit_transform(documents)

# Options
vectorizer = Classifier::TFIDF.new(
  min_df: 2,           # minimum document frequency
  max_df: 0.95,        # maximum document frequency (fraction)
  ngram_range: [1, 2], # unigrams and bigrams
  sublinear_tf: true   # use 1 + log(tf) instead of raw tf
)

# Integrate with existing classifiers
lsi = Classifier::LSI.new(vectorizer: :tfidf)  # Use TF-IDF instead of raw TF
```

## Implementation Notes

- Store IDF values computed during `fit`
- Leverage existing `word_hash` for term frequency extraction
- Support serialization (save/load IDF values)
- Consider sparse vector representation for efficiency
- Optional: n-gram support (bigrams, trigrams)

## Current State

The gem already has:
- `String#word_hash` - term frequency with stemming
- `String#clean_word_hash` - TF with stopword removal

Missing:
- IDF computation across corpus
- Combined TF-IDF weighting
- Vocabulary management (min_df, max_df filtering)

## References

- [TF-IDF Wikipedia](https://en.wikipedia.org/wiki/Tf%E2%80%93idf)
- [scikit-learn TfidfVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html)

## Related

- Enhances Logistic Regression (#102)
- Enhances kNN (#103)  
- Can improve LSI quality

### PR DESCRIPTION
**Pull request by cardmagic**
## Summary

Adds a standalone TF-IDF (Term Frequency-Inverse Document Frequency) vectorizer for text feature extraction.

- **fit/transform/fit_transform API** (scikit-learn style)
- **Vocabulary filtering** via min_df/max_df thresholds
- **N-gram support** (unigrams, bigrams, trigrams)
- **Sublinear TF scaling** (1 + log(tf))
- **L2 normalized** output vectors
- **JSON and Marshal** serialization

Leverages existing `word_hash` infrastructure for term frequency extraction with stemming and stopword removal.

## Example

```ruby
tfidf = Classifier::TFIDF.new(sublinear_tf: true)
tfidf.fit(["Dogs are great pets", "Cats are independent"])
vector = tfidf.transform("Dogs are loyal")
# => {:dog=>0.7071..., :loyal=>0.7071...}
```

## Test plan

- [x] 37 new tests covering all features
- [x] All 327 tests pass
- [x] Rubocop clean
- [x] README updated with TF-IDF section

Closes #104

### CODE REVIEW

**Comment by cardmagic** [comment]
@greptile-apps Thanks for the review! I've addressed the structural issues:

- ✅ Removed "with smoothing" from the IDF comment (line 76)
- ✅ Refactored `validate_df!` to use early returns instead of nested conditionals
- ✅ Inlined intermediate variables in `validate_ngram_range!`

However, I've intentionally kept the method documentation comments (lines 52, 84, 102, 109, 141). These serve as RDoc documentation for the public API and are valuable for users of the gem who rely on generated documentation. The style guide's guidance against redundant comments applies to inline implementation comments, not API documentation.

Could you please re-review when you get a chance?

**Comment by cardmagic** [comment]
@greptile-apps Thanks for the follow-up review! Here's what was addressed:

**Applied:**
- ✅ `extract_terms`: Early return when `ngram_range[1] <= 1`
- ✅ `validate_ngram_range!`: Simplified with `range.all?(Integer) && range.all?(&:positive?)`
- ✅ `within_df_bounds?`: Inlined single-use variables

**Not applied (intentionally kept):**

1. **`validate_df!` (lines 230-235)**: The current code already uses guard clauses - each line raises immediately on invalid input. The suggestion adds early returns for *valid* cases then a conditional raise, which is more complex, not simpler.

2. **`initialize` comments (lines 30-34)**: These explain parameter *semantics*, not just names:
   - `min_df/max_df`: Integer vs Float behavior differs (count vs proportion)
   - `ngram_range`: Example values like `[1,2]` help users
   - `sublinear_tf`: Explains the `1 + log(tf)` formula
   
   This is useful API documentation for gem users, not redundant commentary.

Could you please re-review?

**Comment by cardmagic** [comment]
@greptile-apps One more note on the `with_index` suggestion for the vocabulary indexing (lines 68-74):

This change would introduce a bug. The current manual counter only increments when a term passes the `within_df_bounds?` check. Using `with_index` would assign indices for *all* terms including skipped ones:

**Current (correct):**
```
Term A passes → index 0
Term B skipped → no index  
Term C passes → index 1
```

**With `with_index` (broken):**
```
Term A passes → index 0
Term B skipped → no index
Term C passes → index 2 ❌ (gap in indices)
```

The vocabulary indices need to be contiguous for the vectorizer to work correctly.

Could you please re-review when you get a chance?
